### PR TITLE
exit when allreduce/broadcast error cause timeout

### PR DESCRIPTION
**Pull request by chenqin**
More detail can be found in https://github.com/dmlc/rabit/issues/105

launch async task in background when rabit cluster suffer collective operation failure (socket connection error, other host failure etc). If cluster restore within timeout threshold, async task will return; otherwise it trigger timeout and exit process.

- support rabit_timeout for user to opt-in, by default this feature is closed.
- extensive unit tests with error mocking, run through integration test
- speed up travis ci tests by removing duplicated doc/lint on osx
@hcho3 @trivialfis

### CODE REVIEW

**Comment by chenqin** [comment]
There are some weird osx socket behavior that breaks ci, working on fix

**Comment by chenqin** [comment]
@hcho3 @trivialfis @CodingCat

**Comment by chenqin** [comment]
OSX shows different behavior when we call exit(-1) in different thread. Looking into this.

Linux
```
 ==========] Running 9 tests from 3 test cases.
[----------] Global test environment set-up.
[----------] 3 tests from allreduce_base
[ RUN      ] allreduce_base.init_task
[       OK ] allreduce_base.init_task (0 ms)
[ RUN      ] allreduce_base.init_with_cache_on
[       OK ] allreduce_base.init_with_cache_on (0 ms)
[ RUN      ] allreduce_base.init_with_ring_reduce
[       OK ] allreduce_base.init_with_ring_reduce (0 ms)
[----------] 3 tests from allreduce_base (0 ms total)
[----------] 4 tests from allreduce_robust
[ RUN      ] allreduce_robust.sync_error_timeout
[EXPERIMENTAL] rabit timeout thread expires in 1 second(s)
[       OK ] allreduce_robust.sync_error_timeout (1006 ms)
[ RUN      ] allreduce_robust.sync_error_reset
[EXPERIMENTAL] rabit timeout thread expires in 1 second(s)
[0] rabit timeout thread 140438572955392 starts
[0] rabit timeout task thread 140438572955392 exits
[       OK ] allreduce_robust.sync_error_reset (101 ms)
[ RUN      ] allreduce_robust.sync_success_error_timeout
[EXPERIMENTAL] rabit timeout thread expires in 1 second(s)
[0] rabit timeout thread 139857163970304 starts
[       OK ] allreduce_robust.sync_success_error_timeout (1207 ms)
[ RUN      ] allreduce_robust.sync_success_error_success
[EXPERIMENTAL] rabit timeout thread expires in 1 second(s)
[0] rabit timeout thread 140438572955392 starts
[0] rabit timeout task thread 140438572955392 exits
[       OK ] allreduce_robust.sync_success_error_success (1121 ms)
[----------] 4 tests from allreduce_robust (3435 ms total)
[----------] 2 tests from allreduce_mock
[ RUN      ] allreduce_mock.mock_allreduce
[       OK ] allreduce_mock.mock_allreduce (4 ms)
[ RUN      ] allreduce_mock.mock_broadcast
[       OK ] allreduce_mock.mock_broadcast (4 ms)
[----------] 2 tests from allreduce_mock (8 ms total)
[----------] Global test environment tear-down
[==========] 9 tests from 3 test cases ran. (3444 ms total)
[  PASSED  ] 9 tests.
```
OSX
```
==========] Running 9 tests from 3 test cases.
[----------] Global test environment set-up.
[----------] 3 tests from allreduce_base
[ RUN      ] allreduce_base.init_task
[       OK ] allreduce_base.init_task (0 ms)
[ RUN      ] allreduce_base.init_with_cache_on
[       OK ] allreduce_base.init_with_cache_on (0 ms)
[ RUN      ] allreduce_base.init_with_ring_reduce
[       OK ] allreduce_base.init_with_ring_reduce (0 ms)
[----------] 3 tests from allreduce_base (0 ms total)
[----------] 4 tests from allreduce_robust
[ RUN      ] allreduce_robust.sync_error_timeout
[EXPERIMENTAL] rabit timeout thread expires in 1 second(s)
/Users/travis/build/chenqin/rabit/test/cpp/allreduce_robust_test.cc:34: Failure
Death test: { switch (0) case 0: default: if (const ::testing::AssertionResult gtest_ar = (::testing::internal:: EqHelper<(sizeof(::testing::internal::IsNullLiteralHelper(m.CheckAndRecover(err_type))) == 1)>::Compare("m.CheckAndRecover(err_type)", "false", m.CheckAndRecover(err_type), false))) ; else ::testing::internal::AssertHelper(::testing::TestPartResult::kNonFatalFailure, "/Users/travis/build/chenqin/rabit/test/cpp/allreduce_robust_test.cc", 32, gtest_ar.failure_message()) = ::testing::Message(); std::this_thread::sleep_for(std::chrono::milliseconds(1100)); }
    Result: failed to die.
 Error msg:
[  DEATH   ] 
[  FAILED  ] allreduce_robust.sync_error_timeout (1108 ms)
[ RUN      ] allreduce_robust.sync_error_reset
[EXPERIMENTAL] rabit timeout thread expires in 1 second(s)
[0] rabit timeout thread 123145456910336 starts
[0] rabit timeout task thread 123145456910336 exits
[       OK ] allreduce_robust.sync_error_reset (176 ms)
[ RUN      ] allreduce_robust.sync_success_error_timeout
[EXPERIMENTAL] rabit timeout thread expires in 1 second(s)
[0] rabit timeout thread 123145431621632 starts
/Users/travis/build/chenqin/rabit/test/cpp/allreduce_robust_test.cc:104: Failure
Death test: { switch (0) case 0: default: if (const ::testing::AssertionResult gtest_ar = (::testing::internal:: EqHelper<(sizeof(::testing::internal::IsNullLiteralHelper(m.CheckAndRecover(err_type))) == 1)>::Compare("m.CheckAndRecover(err_type)", "false", m.CheckAndRecover(err_type), false))) ; else ::testing::internal::AssertHelper(::testing::TestPartResult::kNonFatalFailure, "/Users/travis/build/chenqin/rabit/test/cpp/allreduce_robust_test.cc", 102, gtest_ar.failure_message()) = ::testing::Message(); std::this_thread::sleep_for(std::chrono::milliseconds(1200)); }
    Result: failed to die.
 Error msg:
[  DEATH   ] 
[  FAILED  ] allreduce_robust.sync_success_error_timeout (1563 ms)
[ RUN      ] allreduce_robust.sync_success_error_success
[EXPERIMENTAL] rabit timeout thread expires in 1 second(s)
[0] rabit timeout thread 123145456910336 starts
[0] rabit timeout task thread 123145456910336 exits
[       OK ] allreduce_robust.sync_success_error_success (1231 ms)
[----------] 4 tests from allreduce_robust (4078 ms total)
[----------] 2 tests from allreduce_mock
[ RUN      ] allreduce_mock.mock_allreduce
[       OK ] allreduce_mock.mock_allreduce (10 ms)
[ RUN      ] allreduce_mock.mock_broadcast
[       OK ] allreduce_mock.mock_broadcast (9 ms)
[----------] 2 tests from allreduce_mock (19 ms total)
[----------] Global test environment tear-down
[==========] 9 tests from 3 test cases ran. (4098 ms total)
[  PASSED  ] 7 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] allreduce_robust.sync_error_timeout
[  FAILED  ] allreduce_robust.sync_success_error_timeout
 2 FAILED TESTS
```

**Comment by chenqin** [comment]
:build cq$ ./unit_tests
[==========] Running 9 tests from 3 test cases.
[----------] Global test environment set-up.
[----------] 3 tests from allreduce_base
[ RUN      ] allreduce_base.init_task
[       OK ] allreduce_base.init_task (0 ms)
[ RUN      ] allreduce_base.init_with_cache_on
[       OK ] allreduce_base.init_with_cache_on (0 ms)
[ RUN      ] allreduce_base.init_with_ring_reduce
[       OK ] allreduce_base.init_with_ring_reduce (0 ms)
[----------] 3 tests from allreduce_base (1 ms total)

[----------] 4 tests from allreduce_robust
[ RUN      ] allreduce_robust.sync_error_timeout
[EXPERIMENTAL] rabit timeout thread expires in 1 second(s)
[       OK ] allreduce_robust.sync_error_timeout (1033 ms)
[ RUN      ] allreduce_robust.sync_error_reset
[EXPERIMENTAL] rabit timeout thread expires in 1 second(s)
[0] rabit timeout thread 123145318354944 starts
[0] rabit timeout task thread 123145318354944 exits
[       OK ] allreduce_robust.sync_error_reset (203 ms)
[ RUN      ] allreduce_robust.sync_success_error_timeout
[EXPERIMENTAL] rabit timeout thread expires in 1 second(s)
[0] rabit timeout thread 123145344966656 starts
[       OK ] allreduce_robust.sync_success_error_timeout (1251 ms)
[ RUN      ] allreduce_robust.sync_success_error_success
[EXPERIMENTAL] rabit timeout thread expires in 1 second(s)
[0] rabit timeout thread 123145318354944 starts
[0] rabit timeout task thread 123145318354944 exits
[       OK ] allreduce_robust.sync_success_error_success (1129 ms)
[----------] 4 tests from allreduce_robust (3616 ms total)

[----------] 2 tests from allreduce_mock
[ RUN      ] allreduce_mock.mock_allreduce
[       OK ] allreduce_mock.mock_allreduce (14 ms)
[ RUN      ] allreduce_mock.mock_broadcast
[       OK ] allreduce_mock.mock_broadcast (7 ms)
[----------] 2 tests from allreduce_mock (21 ms total)

[----------] Global test environment tear-down
[==========] 9 tests from 3 test cases ran. (3638 ms total)
[  PASSED  ] 9 tests.
cq$ make test
Running tests...
Test project /Users/cq/xgboost/rabit/build
    Start 1: TestRabitLib
1/1 Test #1: TestRabitLib .....................   Passed    3.54 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   3.55 sec

File: .travis.yml [hunk 1]
```diff
@@ -8,16 +8,29 @@ osx_image: xcode10.2
...
+  exclude:
+    - os: osx
+      env: TASK=lint LINT_LANG=cpp
+    - os: osx
+      env: TASK=lint LINT_LANG=python
+    - os: osx
+      env: TASK=doc
+    - os: osx
+      env: TASK=build
+
```
Comment (chenqin) [inline .travis.yml]: speed up travis ci speed

File: CMakeLists.txt [hunk 2]
```diff
@@ -20,12 +20,16 @@ if(R_LIB OR MINGW OR WIN32)
           CXX_STANDARD_REQUIRED ON
           POSITION_INDEPENDENT_CODE ON)
 else()
-  SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pthread")
-  add_library(rabit src/allreduce_base.cc src/allreduce_robust.cc src/engine.cc src/c_api.cc)
-  add_library(rabit_base src/allreduce_base.cc src/engine_base.cc src/c_api.cc)
+  find_package(Threads REQUIRED)
```
Comment (chenqin) [inline CMakeLists.txt]: sidecar timeout async task can potentially runs on different thread.

File: include/rabit/internal/utils.h [hunk 3]
```diff
@@ -86,7 +86,7 @@ inline void HandleAssertError(const char *msg) {
  */
 inline void HandleCheckError(const char *msg) {
   if (STOP_PROCESS_ON_ERROR) {
-    fprintf(stderr, "%s, shutting down process", msg);
+    fprintf(stderr, "%s, shutting down process\n", msg);
```
Comment (chenqin) [inline include/rabit/internal/utils.h]: format fix

**Comment by chenqin** [comment]
@trivialfis can you help review this pr? :)

File: scripts/travis_script.sh [hunk 4]
```diff
@@ -25,9 +25,12 @@ if [ ${TASK} == "cmake-test" ]; then
     mkdir build
     cd build
     cmake -DRABIT_BUILD_TESTS=ON -DRABIT_BUILD_DMLC=ON -DGTEST_ROOT=${HOME}/.local ..
+    # known osx gtest 1.8 issue
+    cp ${HOME}/.local/lib/*.dylib .
     #unit tests
-    make
+    make -j4
```
Comment (trivialfis) [inline scripts/travis_script.sh]: `make -j$(nproc)`

File: scripts/travis_script.sh [hunk 5]
```diff
@@ -25,9 +25,12 @@ if [ ${TASK} == "cmake-test" ]; then
...
     make test
+    #make test VERBOSE=1 || exit -1
```
Comment (trivialfis) [inline scripts/travis_script.sh]: Missed a comment.

File: src/allreduce_base.h [hunk 6]
```diff
@@ -546,6 +546,10 @@ class AllreduceBase : public IEngine {
   int rabit_bootstrap_cache = 0;
   // enable detailed logging
   int rabit_debug = 0;
+  // by default, if rabit worker not recover in half an hour exit
+  int timeout_sec = 1800;
+  // flag to enable rabit_timeout
+  int rabit_timeout = 0;
```
Comment (trivialfis) [inline src/allreduce_base.h]: If it's a flag, then it should be a enum class or bits with documented finite values.  In `all_reduce_base.cc` you write its value with `atoi`, I'm gonna have to assume this is not a simple flag.

File: scripts/travis_script.sh (hunk 4 above)
Comment (chenqin) [reply scripts/travis_script.sh]: fixed.

File: scripts/travis_script.sh (hunk 5 above)
Comment (chenqin) [reply scripts/travis_script.sh]: fixed

File: src/allreduce_base.h (hunk 6 above)
Comment (chenqin) [reply src/allreduce_base.h]: fixed and using bool type, user can set =1 or =true

**Comment by chenqin** [comment]
> I haven't really look into the details yet. But your work is very worth documenting. Could you please start writing user documents?

updated guide doc

File: src/allreduce_robust.cc [hunk 7]
```diff
@@ -55,24 +58,31 @@ bool AllreduceRobust::Shutdown(void) {
   try {
     // need to sync the exec before we shutdown, do a pesudo check point
     // execute checkpoint, note: when checkpoint existing, load will not happen
-    utils::Assert(RecoverExec(NULL, 0, ActionSummary::kCheckPoint, ActionSummary::kSpecialOp,
+    _assert(RecoverExec(NULL, 0, ActionSummary::kCheckPoint, ActionSummary::kSpecialOp,
```
Comment (trivialfis) [inline src/allreduce_robust.cc]: What's the difference?

File: src/allreduce_robust.h [hunk 8]
```diff
@@ -632,6 +633,14 @@ o   *  the input state must exactly one saved state(local state of current node)
...
   // if checkpoint were loaded, used to distinguish results boostrap cache from seqno cache
   bool checkpoint_loaded;
+  // sidecar executing timeout task
+  std::future<bool> rabit_timeout_task;
+  // flag to shutdown rabit_timeout_task before timeout
+  std::atomic<bool> shutdown_timeout{false};
+  // error handler
+  void (* _error)(const char *fmt, ...) = utils::Error;
+  // assert handler
+  void (* _assert)(bool exp, const char *fmt, ...) = utils::Assert;
```
Comment (trivialfis) [inline src/allreduce_robust.h]: I see.  But why?

File: src/allreduce_robust.cc [hunk 9]
```diff
@@ -600,14 +610,44 @@ AllreduceRobust::ReturnType AllreduceRobust::TryResetLinks(void) {
...
+      _error("[%d] exit due to time out %d s\n", rank, timeout_sec);
+      return false;
+    });
+  }
   // simple way, shutdown all links
   for (size_t i = 0; i < all_links.size(); ++i) {
     if (!all_links[i].sock.BadSocket()) all_links[i].sock.Close();
   }
+  // smooth out traffic to tracker
+  std::this_thread::sleep_for(std::chrono::milliseconds(10*rank));
```
Comment (trivialfis) [inline src/allreduce_robust.cc]: Could you explain a little about this line?

**Comment by trivialfis** [comment]
Pretty good overall.  Please see inlined questions.

File: src/allreduce_robust.cc (hunk 9 above)
Comment (chenqin) [reply src/allreduce_robust.cc]: In our production jobs, we observed this were causing tracker failures. It all point to the fact when we have upper handreds of distributed xgboost trainers(a.k.a rabit instances). If one worker dead, the socket OOB msg will propagated to ALL trainer and starts new tcp connections to our python tracker at same time. It often overload tracker and cause entire job crash, at this point, we introduce a 10ms interval to every requests in order to mitigate this case.

File: src/allreduce_robust.cc (hunk 7 above)
Comment (chenqin) [reply src/allreduce_robust.cc]: mocking, please see comments below.

File: src/allreduce_robust.cc (hunk 9 above)
Comment (trivialfis) [reply src/allreduce_robust.cc]: Em.. I will merge it now.  But for long term can you think of a better solution?

File: src/allreduce_robust.h (hunk 8 above)
Comment (trivialfis) [reply src/allreduce_robust.h]: I will jump into this when time allows.

Comment (chenqin) [reply src/allreduce_robust.h]: something like this https://github.com/chenqin/rabit/commit/dba2836eb9c63655de0351d19215b56dc162d0b3 (WIP)

File: src/allreduce_robust.cc (hunk 9 above)
Comment (trivialfis) [reply src/allreduce_robust.cc]: Since this only happens in failure, not in normal workflow, will merge it.

**Comment by trivialfis** [comment]
@chenqin Big thanks for your good work.

**Pull request merged**